app/src/main/java/com/ubergeek42/WeechatAndroid/service/Events.java

```diff
@@ -77,4 +80,54 @@ public static void fireInput(@NonNull Buffer buffer, @Nullable String input) {
             return "SendMessageEvent(message=" + message + ")";
         }
     }
+
+    // https://weechat.org/files/doc/devel/weechat_relay_protocol.en.html#commands
+    // make use of the message identifier so we can correlate the server's response to our message
+    public static class SendMessageWithPromiseEvent {
+        final public String message;
+        final public CompletableFuture<RelayObject[]> future;
+        final public Integer futureId;
+        // The (id) field of the message is also used in BufferList.java for message handlers
+        // e.g. (listbuffers), (nicklist). We may want to add similar handlers for messages
+        // of this type, hence originalMessageId may be used here to indicate the original "purpose"
+        // of the message (see onMessage of RelayService.java)
+        final public String originalMessageId;
+
+        private static int futureCounter = 0;
```

Comment:
we are already doing something similar [here](https://github.com/ubergeek42/weechat-android/blob/df1a9b46923d0450a6055bc35f40f70adc9bdc16/app/src/main/java/com/ubergeek42/WeechatAndroid/relay/BufferList.java#L195-L196), perhaps do it in a single place? these could somehow clash perhaps

Reply:
...not sure how useful `android-retrofuture` would be here. sure we should be doing something better than the very low level approach we are taking now but instead of the future thing you can just do something along `RelayMessageHandler completionhandler = (obj, id) -> { ... runOnMainThread(() -> { ... } };` and you won't need another thread here...

Reply:
As mentioned above, I am doing the counter thing in the same place now with a `RelayMessageHandler`, no more rolling my own counter.

I think Futures are a very elegant solution, because they are cancellable for one, since on a bad / lossy network connection the response might take forever to come, and on top of that we are preventing user input while waiting, so I have a 10 second timeout

https://github.com/ubergeek42/weechat-android/blob/5b0d1d8faf2c9acc1e105f2f2647ef9f4b9396e2/app/src/main/java/com/ubergeek42/WeechatAndroid/tabcomplete/TabCompleter.kt#L106-L121

 Isolating everything in `TabCompleter.kt` would become messy to do if I implemented it solely with a `RelayMessageHandler` because I would have to percolate `tcViaRelayCompletionResults`, `tcViaRelayAddSpace`, `tcViaRelayPosStart`, etc from `BufferList` -> `Buffer` -> `BufferFragment` -> `TabCompleter.kt` or expose `TabCompleter` to `BufferList` somehow which would mean writing a lot more interfaces; I do believe it is a lot more maintainable this way, being able to send the completion request with `buffer.tabCompletion` and handling the result directly in the same block of code

Additionally, i don't think it's a good idea to prevent user input without a guarantee that it will be restored at some point (even if very rarely encountered in practice), which would be very messy to do, imagine writing a 10 second AsyncTask or something to delete the `id` from the `ConcurrentHashMap` or something....